### PR TITLE
Human vs human - two players take turns to play

### PR DESCRIPTION
**Pull request by mercebc**

### CODE REVIEW

File: lib/UI/UI.rb [hunk 1]
```diff
@@ -40,7 +40,7 @@ def show_grid board
         @out.print DIVIDER if (row < size)
         @out.print PLUS if (col < size and row < size)
       end
-      @out.print "\n"
+      @out.print "\n"if (row < size)
```
Comment (ssunday) [inline lib/UI/UI.rb]: Should be a space between the if and what comes before. Also don't need the parenthesis

File: lib/core/player.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+class Player
+  attr_accessor :symbol
+  attr_reader :ui
+
+  def initialize(symbol, ui)
+    @symbol = symbol
+    @ui = ui
+  end
+
+  def move board
```
Comment (ssunday) [inline lib/core/player.rb]: Better readability to have parens around the parameters

File: spec/core/player_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'rspec'
+require 'UI/UI'
+require 'core/player'
+require 'core/board'
+
+describe Player do
+
```
Comment (ssunday) [inline spec/core/player_spec.rb:7]: In rspec, you should use let (https://relishapp.com/rspec/rspec-core/v/2-5/docs/helper-methods/let-and-let) to handle variables across tests.

so let(:board) { Board.new(3) } and so on.

File: spec/core/player_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  before(:each) do
+    @board = Board.new(3)
+    @output = StringIO.new
+  end
+
+  it 'plays her/his turn' do
```
Comment (ssunday) [inline spec/core/player_spec.rb:11]: The only thing I can think with this test is that it has a lot of other class dependencies and it's all testing the output of the UI as well. Is there a way to more directly test whatever player.move is doing?

File: lib/UI/UI.rb [hunk 5]
```diff
@@ -32,31 +27,27 @@ def show_grid board
...
         position += 1
       end
       @out.print "\n"
       for col in 1..size do
-        @out.print DIVIDER if (row < size)
+        @out.print DIVIDER if row < size
         @out.print PLUS if (col < size and row < size)
       end
-      @out.print "\n"
+      @out.print "\n" if row < size
```
Comment (ssunday) [inline lib/UI/UI.rb:38]: This is repeated a few times...maybe extract out?

File: lib/core/flow.rb [hunk 6]
```diff
@@ -17,17 +19,14 @@ def start
...
-    until @board.grid[position].is_empty? do
-    @ui.show_error_message("Please choose a different position where the cell is empty\n")
-    position = @ui.get_position
-    end
-    position
-  end
-
   def play
-    @board.mark(get_valid_position, @ui.get_symbol)
+    position = @players[0].get_valid_position(@board)
```
Comment (ssunday) [inline lib/core/flow.rb]: https://apidock.com/ruby/Array/first ...alternatively, could you make a function that is called current player that does this?

File: lib/core/player.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  attr_reader :ui
+
+  def initialize(symbol, ui)
+    @symbol = symbol
+    @ui = ui
+  end
+
+  def get_valid_position(board)
+    position = @ui.get_position
+    until board.grid[position].is_empty? do
```
Comment (ssunday) [inline lib/core/player.rb]: Maybe extract out into error function?

Comment (mercebc) [reply lib/core/player.rb]: thanks for all the feedback! Im not too sure how to implement an error function. could you send me an example? thanks!!

Comment (ssunday) [reply lib/core/player.rb]: I meant like extract everything out including that until

Comment (ssunday) [reply lib/core/player.rb]: Like this: ```
until board.grid[position].is_empty? do
      @ui.show_error_message("Please choose a different position where the cell is empty\n")
      position = @ui.get_position
 end
```
Hmm. Maybe not break it all out, but it feels like there's a few different things going on that can be pulled out. The error clause, showing the error message, etc.

Comment (mercebc) [reply lib/core/player.rb]: thanks!

File: lib/core/flow.rb [hunk 8]
```diff
@@ -20,13 +20,18 @@ def start
...
     @ui.show_grid(@board)
     swap_players
   end
 
   def swap_players
     @players[0], @players[1] = @players[1], @players[0]
   end
+
+  def current_player
+    @players.first
```
Comment (ssunday) [inline lib/core/flow.rb:34]: 😁

File: lib/core/flow.rb [hunk 9]
```diff
@@ -20,13 +20,18 @@ def start
...
-    position = @players[0].get_valid_position(@board)
-    @board.mark(position, @players[0].symbol)
+    position = current_player.get_valid_position(@board)
+    @board.mark(position, current_player.symbol)
     @ui.show_grid(@board)
     swap_players
   end
 
   def swap_players
     @players[0], @players[1] = @players[1], @players[0]
```
Comment (ssunday) [inline lib/core/flow.rb]: Could you implement the first/last here?

File: lib/core/player.rb [hunk 10]
```diff
@@ -11,12 +11,17 @@ def initialize(symbol, ui)
...
-    until board.grid[position].is_empty? do
-      @ui.show_error_message("Please choose a different position where the cell is empty\n")
-      position = @ui.get_position
-    end
-    position
+    validate_position(@ui.get_position, board)
   end
 
+  def validate_position(position, board)
+    board.invalid_cell(position) ? get_validated_position(board) : position
```
Comment (ssunday) [inline lib/core/player.rb:18]: Woot!

File: lib/core/player.rb [hunk 11]
```diff
@@ -11,12 +11,17 @@ def initialize(symbol, ui)
...
+  end
+
+  def get_validated_position(board)
+    @ui.show_error_message("Please choose a different position where the cell is empty\n")
+    get_valid_position(board)
+  end
+
+  private :validate_position, :get_validated_position
```
Comment (ssunday) [inline lib/core/player.rb:26]: You can have the `private` and everything below it is private

File: spec/core/board_spec.rb [hunk 12]
```diff
@@ -37,4 +37,9 @@
     end
     expect(board.is_full?).to eq(true)
   end
+
+  it 'a cell that is not empty is invalid' do
```
Comment (ssunday) [inline spec/core/board_spec.rb:41]: Could you write the test when it's false?

**Pull request merged**